### ISSUE TITLE
Remove ETag from the digest key calculations

### ISSUE THREAD

**Issue by yoavweiss** [httpwg/http-extensions#516]
Having ETags as part of the digest key adds a lot of complexity to the Cache Digest implementation both on the client and the server:
* Clients may not have the ETag readily available at removal time, which would mean removal from the digest introduces I/O operations and slowness.
* Servers may not have the ETag readily available when they want to send down push promises, forcing them to make those conditional on I/O operations (and again, slowness).

I [ran an HTTPArchive analysis](https://discuss.httparchive.org/t/cacheability-of-render-critical-resources/1278) and it seems that from the render-critical resources that are push candidates, 77.26% of them are cacheable for at least 24 hours.
Same query with 30 days minimal freshness gives us 52.32% of the resources.

So it seems that servers can in many cases have enough render-critical resources to push even without the ETag info.
Therefore I suggest removing it, to simplify implementations. If we'd see that it becomes an issue in the future, we could add it back as optional.

Thoughts?

**Comment by kazuho** [comment]
> So it seems that servers can in many cases have enough render-critical resources to push even without the ETag info.

I agree.

Note also that it is difficult to use a digest of freshly cached responses that uses ETag due to lack of agreement on if we can push a response to replace a freshly cached object within the client's cache. See the chart on https://datatracker.ietf.org/meeting/101/materials/slides-101-httpbis-cache-digest-00 page 7.

> Therefore I suggest removing it, to simplify implementations. If we'd see that it becomes an issue in the future, we could add it back as optional.

The reason we allow the use of ETag is to support sending a digest of stale responses. When a server receives a digest of stale responses, a server needs to decide if it pushes a 200 or "not modified", based on the value of ETag.

In other words, removal of ETag support needs to come along with removing support for stale digests. That said, I do not oppose to removing them altogether (means that we will only allow sending a digest of freshly-cached responses that uses HASH(URL) as the key), considering the fact that how to push a "not modified" signal is not something that the community has agreed.

**Comment by evert** [comment]
I'm a little sad about this, so I feel it's important to explain my use-case for h2 push + cache digests.

I'm mainly concerned with building REST API's. Traditionally when building API's there's often been a need to be able to express a 'collection' and an 'item'.

A big reason for needing this, is that we can't ask the client to do a GET request for every individual item in a collection, so they get bunched together into some kind of collection resource.

However, those collection resources still contain some kind of representation of the item in the collection, and can often indicate to the client what real item they refer to.

A real example of this is [HAL](http://stateless.co/hal_specification.html). If you're unfamiliar with HAL, it actually expresses collections though an "_embedded" format. It almost literally expresses:

> "Here is a collection. To save you time and bandwidth, we're embedding the representations of all the resources that are contained within the collection"

Many formats will have some variation of this idea. But the idea is not perfect. Specifically:

* HTTP caches will not be aware or understand embedded items, and will not invalidate embedded items if they got a new copy.
* It's often only possible to embed items that are identical to the parent format (in my case HAL).
* It's often possible to include additional request/response headers.

The more I've been thinking about this and generic clients, I feel that HTTP/2 push is a great solution for this. A conforming client can just pretend do GET requests that come from the push-cache. I've started testing with this, and conceptually I think it works really well.

This could all be done without cache-digest, but the ability to only push resource states that have changed is just such a powerful idea to me.

However, in my mind ETag is the right mechanism for this. I think the statistics you cited are right, but HTTP/2 push and Cache-Digests can be an enabler for this type of technology. Rarely APIs will know in advance how long resources should be cacheable, because changes are usually based on user actions.

**Comment by evert** [comment]
Thank you @kazuho , I appreciate!
